reusable_components_nlp/README.md

```diff
@@ -0,0 +1,989 @@
+            html_unescaped =Transformer.\
+                __html_parser.unescape(decoded) 
+            html_unescaped = re.sub(r"\r\n", " ", html_unescaped)
+            html_unescaped = re.sub(r"\r\r\n", " ", html_unescaped)
+            html_unescaped = re.sub(r"\r", " ", html_unescaped)
+            html_unescaped = html_unescaped.replace("&gt;", " > ")
+            html_unescaped = html_unescaped.replace("&lt;", " < ")
+            html_unescaped = html_unescaped.replace("--", " - ")
+            html_unescaped = Transformer.__uplus_pattern.sub(
+                " U\g<digit> ", html_unescaped)
+            html_unescaped = Transformer.__markup_link_pattern.sub(
+                " \1 \2 ", html_unescaped)
+            html_unescaped = html_unescaped.replace("\\", "")
+            return html_unescaped
+    
+
+
+If you want to understand how the CLI pipeline talks to each other, have a look at the end to end test in `pipeline/pipeline_tests/`:
+
+
+```python
+!pytest ./pipeline/pipeline_tests/. --disable-pytest-warnings
+```
+
+    [1m============================= test session starts ==============================[0m
+    platform linux -- Python 3.7.3, pytest-4.5.0, py-1.8.0, pluggy-0.11.0
+    rootdir: /home/alejandro/Programming/kubernetes/seldon/seldon-core/examples/kubeflow
+    collected 1 item                                                               [0m[1m
+    
+    pipeline/pipeline_tests/test_pipeline.py [32m.[0m[36m                               [100%][0m
+    
+    [33m[1m===================== 1 passed, 9 warnings in 3.26 seconds =====================[0m
+
+
+To build the image we provide a build script in each of the steps that contains the instructions:
+
+
+```python
+!cat pipeline/pipeline_steps/clean_text/build_image.sh
+```
+
+    #!/bin/bash
+    
+    s2i build . seldonio/seldon-core-s2i-python3:0.6 clean_text_transformer:0.1
+    
+
+
+The only thing you need to make sure is that Seldon knows how to wrap the right model and file.
+
+This can be achieved with the s2i/environment file. 
+
+As you can see, here we just tell it we want it to use our `Transformer.py` file:
+
+
+
+```python
+!cat pipeline/pipeline_steps/clean_text/.s2i/environment
+```
+
+    MODEL_NAME=Transformer
+    API_TYPE=REST
+    SERVICE_TYPE=MODEL
+    PERSISTENCE=0
+
+
+Once this is defined, the only thing we need to do is to run the `build_image.sh` for all the reusable components.
+
+Here we show the manual way to do it:
+
+
+```bash
+%%bash
+# we must be in the same directory
+cd pipeline/pipeline_steps/clean_text/ && ./build_image.sh
+cd ../data_downloader && ./build_image.sh
+cd ../lr_text_classifier && ./build_image.sh
+cd ../spacy_tokenize && ./build_image.sh
+cd ../tfidf_vectorizer && ./build_image.sh
+```
+
+## 3) Train our NLP Pipeline through the Kubeflow UI
+We can access the Kubeflow dashboard to train our ML pipeline via http://localhost/_/pipeline-dashboard
```

Comment:
How about instructing users to connect to their dashboard in the usual way with an external URL or giving them a port-forward command?
`kubectl -n kubeflow port-forward svc/ambassador 8080:80`

Reply:
Fair enough